File: Java/src/PathSumII.java
import java.util.ArrayList;
import java.util.List;

public class PathSumII {
	
	public List<List<Integer>> pathSum(TreeNode root, int sum) {
		ArrayList<List<Integer>> result = new ArrayList<List<Integer>>();
        ArrayList<Integer> currentList = new ArrayList<Integer>();
		
		pathSumHelper(root, sum, 0, result, currentList);
        
        return result;
    }
	
	private void pathSumHelper(TreeNode root, int sum, int currentSum, 
			ArrayList<List<Integer>> result, ArrayList<Integer> currentList){
		if(root == null) return;
		
		currentSum += root.val;
		if(root.left == null && root.right == null && currentSum == sum) {
			ArrayList<Integer> copy = new ArrayList<Integer>(currentList);
			copy.add(root.val);
			result.add(copy);
			return;
		}
		
		currentList.add(root.val);
		pathSumHelper(root.left, sum, currentSum, result, currentList);
		pathSumHelper(root.right, sum, currentSum, result, currentList);
		currentList.remove(currentList.size() - 1);
		
	}
    
    

}

File: Python/PathSumII.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from TreeNode import TreeNode

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        self.tar = sum
        self.res = []
        self.dfs(root, [], 0)
        return self.res
        

    def dfs(self, n, l, sum):
        if n == None:
            return
        if n.left == None and n.right == None:
            if sum + n.val == self.tar:
                copy = list(l)
                copy.append(n.val)
                self.res.append(copy)
        
        l.append(n.val)
        self.dfs(n.left, l, sum + n.val)
        self.dfs(n.right, l, sum + n.val)
        l.pop(len(l)-1)

if __name__ == "__main__":
    test = Solution()
    root = TreeNode(1)
    res = test.pathSum(root, 1)
    print(res)

